### PR TITLE
Alternately updated DAN

### PR DESCRIPTION
**Pull request by akuzeee**
- E, M, DのうちDに関してのみ勾配を求められるようにDANクラスのインスタンスメソッド(pred_d)を追加

### CODE REVIEW

**Comment by akuzeee** [comment]
## MNISTRのスコア

### DAN(ReverseLayerF), best test score

| target | best_test_acc | 
| ---: | ---: |
| 0  | 0.841 |
| 15 | 0.986 |
| 30 | 0.980 |
| 45 | 0.930 |
| 60 | 0.977 |
| 75 | 0.862 |


### DAN(alt), best test score
| target | best_test_acc | 
| ---: | ---: |
| 0  | 0.836 |
| 15 | 0.985 |
| 30 | 0.973 |
| 45 | 0.942 |
| 60 | 0.978 |
| 75 | 0.864 |

File: examples/MNISTR_DAN_alt.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+                    loss = y_loss + d_loss
+                    loss.backward()
+                    y_optimizer.step()
+                    y_acc = calc_acc(y_pred, y)
+                    y_losses.append(y_loss.data[0])
+                    y_acces.append(y_acc)
+
+                # update Domain Discriminator
+                else:
+                    d_optimizer.zero_grad()
```
Comment (yusuke0519) [inline examples/MNISTR_DAN_alt.py]: @akuzeee 普通Dを先に学習する気がします。

File: pytransfer/learners/dan.py [hunk 2]
```diff
@@ -43,6 +43,13 @@ def forward(self, input_data):
         domain_output = self.D(reverse_feature)
         return class_output, domain_output
 
+    def pred_d(self, input_data):
+        input_data.requires_grad = False
+        feat = self.E(input_data)
+        feat = Variable(feat.data, requires_grad=True)
```
Comment (yusuke0519) [inline pytransfer/learners/dan.py]: 一回dataにしてしまえば勾配それより戻らないということですね。勉強になりました。

File: examples/MNISTR_DAN.py [hunk 3]
```diff
@@ -159,12 +230,19 @@ def main():
         M = Classifier(NUM_CLASSES, E.output_shape())
         D = Discriminator(len(domains) - 1, E.output_shape())
         learner = DAN(E, M, D, 0.1).cuda()
```
Comment (yusuke0519) [inline examples/MNISTR_DAN.py]: ここ、伝わってるかわからないので再度ですが、annealingに関するブランチをマージしてる場合挙動が変わってます（デフォルトでアニーリングオフになってます）。

Comment (akuzeee) [reply examples/MNISTR_DAN.py]: Annealingの挙動ですが，気になったところコメント返したので確認してもらってよいですか？
https://github.com/yusuke0519/TL_utils/pull/20

**Comment by akuzeee** [comment]
メモ: MNISTR_CNN, 訓練の度にスコアが変わるのが気持ち悪い．モデルの初期化の値と，サンプラーはseedで固定できていることを確認

**Comment by yusuke0519** [comment]
1. annealingに関する部分マージして修正しました
2. foward部分、少しインタフェースを修正したんでコメントもらえると助かります。
具体的には、pred_dをする代わりに、fowardに
freeze_E：Encoderの勾配を利用するかどうか
use_reverse_layer：GradientReversalLayerを利用するかどうか
require_class：クラスラベルを利用するか（Mを通すかどうか）
の3つの引数追加しています。

altanativeに訓練する場合、gradient reverseをしない方法の実装もあり得る（GANみたいにラベル自体を反転する）のでこんな形にしてます。
